File: py/merge_sort.py
import math
import sys
import time


if __name__ == "__main__":
	
	direc = sys.argv[1]

	with open(f"./data/{direc}.txt", "r") as f:
		txt = f.read()
	
	if not txt:
		txt = input()

	array = [int(x.strip()) for x in txt.split(',')]

	array_n = len(array)

	# algo start
	def merge(arr_1, arr_2):
		ans_arr = []
		i1, i2 = 0, 0

		while i1 < len(arr_1) and i2 < len(arr_2):
			if arr_1[i1] < arr_2[i2]:
				ans_arr.append(arr_1[i1])
				i1 += 1
			else:
				ans_arr.append(arr_2[i2])
				i2 += 1

		if i1 == len(arr_1):
			ans_arr += arr_2[i2:]
		else:
			ans_arr += arr_1[i1:]
		return ans_arr

	def merge_sort(array):

		len_a = len(array)

		if len_a == 1:
			return array 
		elif len_a == 2:
			if array[1] < array[0]:
				array = [array[1], array[0]]
			return array 

		half_n = math.floor(len_a / 2)

		return merge(merge_sort(array[:half_n]), merge_sort(array[half_n:]))

	_start = time.time()
	array = merge_sort(array)
	_end = time.time()

	with open(f"./result/py.txt", "a") as f:
		f.write(f"merge_sort, {direc}, {_end - _start}\n")

File: js/merge_sort.js
// input dependencies
const prompt = require('readline-sync');
const fs = require('fs');

let direc = process.argv.slice(2)[0];
let txt = fs.readFileSync(`./data/${direc}.txt`).toString();

if (txt.length === 0) {
	txt = prompt.question();
}

let array = txt.split(',').map((a) => {
	return parseInt(a.trim(), 10);
});

function main() {

	// algo start
	const merge = function (arr1, arr2) {
		let ans_arr = [];
		let ir = 0, il = 0;

		while (ir < arr1.length && il < arr2.length) {
			if (arr1[ir] < arr2[il]) {
				ans_arr.push(arr1[ir]);
				ir ++;
			} else {
				ans_arr.push(arr2[il]);
				il ++;
			}
		}

		if (arr1.length == ir) {
			ans_arr = ans_arr.concat(arr2.slice(il));
		} else {
			ans_arr = ans_arr.concat(arr1.slice(ir));
		}
		return ans_arr;
	}

	const mergesort = function (arr) {
		let len_arr = arr.length;

		if (len_arr == 1) {
			return arr;
		} else if (len_arr == 2) {
			if (arr[0] > arr[1]) {
				arr = Array(arr[1], arr[0]);
			}
			return arr;
		} else {
			let half_arr = Math.floor(len_arr / 2);

			return merge(mergesort(arr.slice(0, half_arr)), mergesort(arr.slice(half_arr)));
		}
	}

	_start = process.hrtime();
	array = mergesort(array);
	_end = process.hrtime();
}

main();

fs.writeFileSync("./result/js.txt", `merge_sort, ${direc}, ${(_end[0] + _end[1]/1000000000) - (_start[0] + _start[1]/1000000000)}\n`, {encoding: 'utf8', flag:'a'});